make-darwin-bundle: Use actual bin output

toonn opened the pull request:
The script and the hook assume `/bin` is in `$out` but that's not always
true for a multi-output derivation.

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

This is very similar to #163516 but for Darwin application bundles.
Nothing uses these hooks yet in Nixpkgs as far as I can tell so this change should be fairly low impact.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

toonn committed:
make-darwin-bundle: Use actual bin output

The script and the hook assume `/bin` is in `$out` but that's not always
true for a multi-output derivation.

toonn committed:
desktopToDarwinBundle: Fixup Exec

The "Exec" key in desktop items sometimes has one of the `%f`, `%F`,
`%u` and `%U` suffixes, which specify whether the command takes a file,
multiple files or a generalized URL or URLs. Darwin application bundles
do no understand this syntax so we do the next best thing, which is
simply dropping it.

milahu commented on pkgs/build-support/setup-hooks/desktop-to-darwin-bundle.sh:
```diff
@@ -150,16 +150,16 @@ convertDesktopFile() {
     local -r sharePath=$(dirname "$(dirname "$file")")
     local -r name=$(getDesktopParam "${file}" "^Name")
-    local -r exec=$(getDesktopParam "${file}" "Exec")
+    local -r exec=$(getDesktopParam "${file}" "Exec" | sed -e 's/ %[fFuU]$//')
```
why this sed?

toonn replied:
It's in the commit message:
> The "Exec" key in desktop items sometimes has one of the `%f`, `%F`,
`%u` and `%U` suffixes, which specify whether the command takes a file,
multiple files or a generalized URL or URLs. Darwin application bundles
do no understand this syntax so we do the next best thing, which is
simply dropping it.

toonn replied:
Yeah, that may be a good idea.
Slightly tangent to the work here though, which is about getting it to actually work when first : )

toonn replied:
What would I use to log from a setup hook?

milahu replied:
echo : )

something like

```
    ${if customExec != null then ''
      local -r exec=${lib.escapeShellArg customExec}
    '' else ''
      local -r execRaw=$(getDesktopParam "${file}" "Exec")
      local -r exec=$(echo "$execRaw" | sed -e 's/ %[fFuU]$//')
      [[ "$exec" != "$execRaw" ]] && echo "makeDarwinBundle: patching exec from '$execRaw' to '$exec'. maybe override with customExec"
    ''}
```

toonn replied:
I added a new custom desktop entry key, like for the icon squircling. It has to be entirely shell since this isn't in a Nix expression.

toonn commented:
The failure on CI does seem valid to me. If anyone knows when and where `outputBin` is available that'd be a great help.

milahu commented on pkgs/build-support/make-darwin-bundle/default.nix:
```diff
@@ -12,14 +12,14 @@
   function makeDarwinBundlePhase() {
-    mkdir -p "$out/Applications/${name}.app/Contents/MacOS"
-    mkdir -p "$out/Applications/${name}.app/Contents/Resources"
+    mkdir -p "${!outputBin}/Applications/${name}.app/Contents/MacOS"
```
you must escape outputBin

see https://grep.app/search?q=outputBin&filter[repo][0]=NixOS/nixpkgs&filter[lang][0]=Nix&filter[path][0]=pkgs/

toonn replied:
Ah, right. Thanks! 🤦

toonn committed:
make-darwin-bundle: Escape outputBin for Nix '' string

toonn committed:
desktopToDarwinBundle: Complete field code removal

Checked the desktop entry spec, there's other field codes than `%[fFuU]`
and those can in fact occur more than once, hence dropping '$' and
adding `/g`.

toonn committed:
desktopToDarwinBundle: Add X-macOS-Exec and log editing Exec

Co-authored-by: milahu <milahu@gmail.com>

milahu commented on pkgs/build-support/make-darwin-bundle/default.nix:
```diff
@@ -12,14 +12,16 @@
 
     if [ -n "${icon}" ]; then
-      ln -s "${icon}" "$out/Applications/${name}.app/Contents/Resources"
+      ln -s "${icon}" \
+        "''${!outputBin}/Applications/${name}.app/Contents/Resources"
```
pls dont wrap line

milahu commented on pkgs/build-support/make-darwin-bundle/default.nix:
```diff
@@ -12,14 +12,16 @@
     fi
 
-    ${writeDarwinBundle}/bin/write-darwin-bundle "$out" "${name}" "${exec}"
+    ${writeDarwinBundle}/bin/write-darwin-bundle "''${!outputBin}" "${name}" \
+      "${exec}"
```
pls dont wrap line

toonn committed:
make-darwin-bundle: Prefer long lines to splitting

toonn commented:
Closing in favor of  #165111, which is cleaner.
